Update crypto packages

vivianjeng opened the pull request:
<!-- Please refer to our contributing documentation for any questions on submitting a pull request -->

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

<!-- Please do not submit updates to dependencies unless it fixes an issue. -->

<!-- Please try to limit your pull request to one type, submit multiple pull requests if needed. -->

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [x] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
- crypto package depends on `circomlibjs` and `maci-crypto` but most of the crypto functions are not used in unirep
- `IncrementalMerkleTree` and `SparseMerkleTree` use different method names

Issue Number: N/A

## What is the new behavior?


<!-- Please describe the behavior or changes that are being added by this PR. -->

- Write crypto functions that are used in Unirep protocol but importing the packages
- Let `0` be the default leaf value in `IncrementalMerkleTree`
- import `poseidon` code from `circomlibjs`
- Use the same method names in `IncrementalMerkleTree` and `SparseMerkleTree`
  (`tree.root`, `tree.createProof()`, `tree.verifyProof()`)
- Write down crypto test scripts and update README.md
- Modify all other packages that depend on `@unirep/circuits`

## Does this introduce a breaking change?

- [x] Yes
- [ ] No

<!-- If this introduces a breaking change, please describe the impact and migration path for existing applications below. -->

## Other information

<!-- Any other information that is important to this PR such as screenshots of how the component looks before and after the change. -->

vimwitch commented on packages/crypto/src/IncrementalMerkleTree.ts:
```diff
@@ -10,7 +16,9 @@ export class IncrementalMerkleTree extends zkIncrementalMerkleTree {
      * @param zeroValue Zero values for zeroes.
      * @param arity The number of children for each node.
      */
-    constructor(depth: number, zeroValue: any, arity?: number) {
```
This export is a a monkey patch on top of `zk-kit`. Instead of doing this I think it would be better to remove this export entirely and have downstream code depend directly on `zk-kit`.

vivianjeng replied:
yeah crypto doesn't have to maintain this. Just to make downsteam easier to use

vimwitch commented on packages/crypto/src/SparseMerkleTree.ts:
```diff
@@ -22,44 +20,38 @@ const newWrappedPoseidonT3Hash = (
 
-    constructor(protected db, private height: number) {
-        assert(height > 0, 'SMT height needs to be > 0')
+    constructor(protected db, private _height: number, zeroHash: bigint) {
```
I'm not sure what `db` argument is here. I know that a `keyv` instance is passed here but not much else. I don't like `keyv` because it doesn't compile well for browsers, it has a lot of nodejs specific requirements.

Could we instead use the `zk-kit` `SparseMerkleTree` implementation?

vivianjeng replied:
Their implementation is really different. I discussed with them but haven't found a solution to use the package. So I just kept the old one.

yeah we can remove `keyv`

vimwitch commented on packages/crypto/src/crypto.ts:
```diff
@@ -1,18 +1,103 @@
+
+import { poseidon } from './poseidon'
+
+// Copy from maci-crypto@0.9.1
```
Why are we copying the `maci-crypto` code here? I think re-exporting it to simplify use is okay, but copying it means we don't get any upstream patches or fixes.

vivianjeng replied:
Because it will download all dependencies we don't need it while installation. 😅
I want to decrease the dependencies and we can only maintain what we need

vimwitch commented on packages/crypto/src/crypto.ts:
```diff
@@ -1,18 +1,103 @@
+ * Compute a random BigInt within SNARK_FIELD_SIZE
+ * @return A random BigInt salt.
+ */
+const genRandomNumber = (): bigint => {
```
Really don't like this rename. Is this implementation any different than `genRandomSalt`?

vivianjeng replied:
actually no. Yeah we can use the old name

vimwitch commented on packages/crypto/src/poseidon.ts:
```diff
@@ -0,0 +1,87 @@
+import assert from 'assert'
```
Where did this file come from/what are we using it for?

vivianjeng replied:
it is from `circomlibjs@0.0.8` 
also don't want to depend on too many packages we don't need 😅

vimwitch commented on packages/crypto/test/IncrementalMerkleTree.test.ts:
```diff
@@ -0,0 +1,38 @@
+    let tree: IncrementalMerkleTree
+    const depth = 4
+
+    it('constructor', async () => {
```
We use the `zk-kit` implementation for this, we shouldn't test it ourselves. Instead our higher level tests should fail if the tree implementation is broken.

vimwitch reviewed (changes requested):
From what I can tell this PR mostly changes some functions.

crypto - `genRandomSalt()` -> `genRandomNum()`
crypto - `IncrementalMerkleTree` constructor arguments
crypto - `SparseMerkleTree.getRootHash()` -> `SparseMerkleTree.root`
smt - `getMerkleProof` -> `createProof`

types - `BigInt` -> `bigint`

I think these changes are more about preference than need. Changing function names and types break any code that depends on this. We should avoid making these types of changes until we have to, and then we should plan and implement it in one go in all of our packages.

The question I'm asking myself here is "are these changes worth the cost?"

vimwitch commented:
I was also thinking a few days ago that we should almost get rid of the crypto package. We're mostly re-exporting other implementations (which is a good thing). I think it's okay to curate things a bit so people don't have to pull `zk-kit` and `maci-crypto` and other stuff, but we should avoid monkey patching the exports.

vivianjeng commented:
New changes has been made:
1. revert `genRandomNumber` to `genRandomSalt`
2. remove `db` as input in `SparseMerkleTree`. but it will be more easily to run out of memory.
    then `SparseMerkleTree` related functions don't need to be `async`

And about `maci-crypto` and `circomlibjs` dependencies, I think we don't need to depend on them to install what we don't need. How do you think? (I am also okay to move the packages back😂)
(I think maybe we (and PSE team) can create a more dependencies-free package and we can import it like `zk-kit/poseidon` without installing other irrelevant packages.)

vimwitch commented:
Dependency management with nodejs is a bit of a nightmare. We need to look at some actual numbers when we make these decisions. On the `master` branch a clean install with no yarn lock results in `1005M` of `node_modules`. On this branch a clean install makes a `1.1G` `node_modules`. I'm not sure how but the package changes increased the size.

We really want to focus on reducing 2 things, production dependency size (when running servers that depend on Unirep) and the bundle size (in browser). On servers the dependencies can be a bit bigger because the much larger item with be the database file. To help here we should make sure that our `dependencies` and `devDependencies` are accurate. For browsers we want to make sure that we don't bundle multiple versions of packages, and that packages we use are tree-shakeable. For `maci-crypto` they export each variable separately, so anything we don't import will be removed from the bundle automatically. The same is true for `circomlibjs`, the output bundle size should be the same if we copy the code like in this pr or if we import normally. So I think it's fine to revert and just import.

Looking at the dependencies closer, even though we removed `keyv` from our packages we still transitively depend on `web3` which depends on `keyv` D: Still good though because it's not being bundled now. On server the best we can do is make sure that `yarn --production` works (for example this will exclude `keyv` because it's `@ethereum-waffle`->`ganache-core`->`web3`->`keyv`)

vivianjeng commented:
> I'm not sure how but the package changes increased the size.

Sorry I also some git commit tools (but I removed now). It can be in another PR.
It is 891M now

> On servers the dependencies can be a bit bigger because the much larger item with be the database file.

Yeah maybe I worried too much.

I reverted some changes, so the PR now is 
1. Set `defaultLeaf=0` and `arity=2` in `IncrementalMerkleTree`
2. Use `poseidon` from `circomlibjs`
3. Remove `keyv` from `SparseMerkleTree` and its methods are non-async functions
4. Update packages and dependencies
5. Use the same method names in `IncrementalMerkleTree` and `SparseMerkleTree`
(tree.root, tree.createProof(), tree.verifyProof())
6. Add crypto CI
7. Update README.md

vimwitch commented on packages/crypto/src/identity.ts:
```diff
@@ -1,13 +1,7 @@
-import _sha256 from 'crypto-js/sha256'
-import { poseidon } from 'circomlibjs'
+import { createHash } from 'crypto'
+const { poseidon } = require('circomlibjs')
```
`require` -> `import`

vimwitch reviewed (approved):
LGTM, just need to switch `require` to `import` in a couple places. Let's make sure to squash the commits during merge
